build: add support for LTO and PGO to the building system

tchaikov opened the pull request:
This changeset ports LTO and PGO support from scylla-enterprise.git to scylladb.git. 

Add support for Link-Time Optimization (LTO) and Profile-Guided Optimization (PGO) 
to improve performance. LTO provides ~7% performance gain and enables crucial 
binary layout optimizations for PGO.

LTO Changes:
- Add `-flto` flag to compile and link steps
- Use `-ffat-lto-objects` to generate both LLVM IR and machine code
- Enable cross-object optimization while maintaining fast test linking

PGO Implementation:
- Implement three-stage build process:
  1. Context-free profiling (`-fprofile-generate`)
  2. Context-sensitive profiling (`-fprofile-use` + `-fcs-profile-generate`)
  3. Final optimization using merged profiles
- Add release-pgo and release-cs-pgo build stages
- Integrate with ninja build system
- Stages can be enabled independently

Profile Management:
- Add `pgo/pgo.py` for workload profile collection
- Store default profile in `pgo/profiles/profile.profdata.xz` using Git LFS
- Add configure.py integration for profile detection and validation
- Support custom profiles via `--use-profile` flag
- Add profile regeneration script

Both optimizations are recommended for maximum performance, though each PGO 
stage adds a full build cycle. Future optimization may allow dropping one 
PGO stage if performance impact is minimal.

---

this is a forward port, hence no need to backport.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14058/)
#### Build Details:
- Duration: 16 min
- Builder: i-0fb5fd8ec533409e8 (m5ad.8xlarge)

tchaikov commented:
v2:

- restructure the commits to squash the fixes into the commit fixed by it.
- add quotes around expressions like `"$(dirname $(realpath $0))"` to prevent word splitting in bash scripts
- add `&` before `2` in
    ```sh
    echo "Error: $SCRIPT_PATH should be ran with $PROJECT_BASE instead of $WORKING_DIR as the working directory" >2
    ```

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14074/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5629/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2515/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/885/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13642/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/289/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14074/)
#### Build Details:
- Duration: 3 hr 59 min
- Builder: spider3.cloudius-systems.com

avikivity commented:
Please adjust the cover letter to be a normal feature introduction, and check again with diff that it is complete.

tchaikov commented:
v3:

- updated the cover letter to be a normal feature introduction
- compared `configure.py`, related CMake scripts and `pgo` subdirectory with enterprise branch, added the missing changes:

  * added `profile_deps` dependency to `build {obj}` rule.
  * corrected the "description" of `xz_uncompress` rule

tchaikov commented:
> Please adjust the cover letter to be a normal feature introduction, and check again with diff that it is complete.

done and done.

tchaikov commented:
@avikivity Hi Avi, when comparing `configure.py`, i noticed that you changed the "optimization-level" of "debug" build from "g" to "0" in https://github.com/scylladb/scylla-enterprise/commit/d0a60b1d6ffc5f73441fa2723ef89e478be18e2e#diff-4d5f3192809ec1b9add6b33007e0c50031ad9a0a2f3f55a481b506468824db2cL377 . this change does not exist in scylladb/scylladb master. and i failed to find related explanation in the commit message of d0a60b1d6ffc5f73441fa2723ef89e478be18e2e . shall we port this change back to this repo? if yes, could you please provide some context of this change?

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14088/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/894/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13650/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2523/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5637/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14088/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/298/)
#### Build Details:
- Duration: 4 hr 22 min
- Builder: spider7.cloudius-systems.com

mykaul commented on pgo/conf/clustering.yaml:
```diff
@@ -0,0 +1,92 @@
+# The CQL for creating a keyspace (optional if it already exists)
+#
+keyspace_definition: |
+  CREATE KEYSPACE keyspace1 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1};
```
unrelated to this PR - I wonder what code doesn't get LTO/PGO treatment by the fact we are using RF=1, which is not the common RF.

mykaul replied:
I see below we have other RFs.

raphaelsc commented on cmake/enable_lto.cmake:
```diff
@@ -0,0 +1,30 @@
+
+#
+# SPDX-License-Identifier: ScyllaDB-Proprietary
+#
```
LicenseRef-ScyllaDB-Source-Available-1.0 ?

mykaul replied:
I wonder if we can add a license check as a GH action.

raphaelsc replied:
> I wonder if we can add a license check as a GH action.

That'd be great. /cc @yaronkaikov

tchaikov replied:
> I wonder if we can add a license check as a GH action.

created at #22065

tchaikov replied:
thank you Raphael for pointing this out. should be fixed in the latest revision.

tchaikov commented:
v4:

- replace two "ScyllaDB-Proprietary" SPDX IDs with the new "LicenseRef-ScyllaDB-Source-Available-1.0" ID.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14103/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13659/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2532/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5646/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/306/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/902/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14103/)
#### Build Details:
- Duration: 4 hr 22 min
- Builder: spider8.cloudius-systems.com

michoecho reviewed (approved):
I did a shallow review, looks fine to me. If there are any problems, we will probably learn about them soon enough.

tchaikov commented:
thank you Michał, for your review and approval.

tchaikov commented:
@scylladb/scylla-maint hello maintainers, could you help queue this change?

avikivity commented:
The subject is still wrong.

tchaikov commented:
> The subject is still wrong.

@avikivity the subject of this PR should be better now. could you take another look?

avikivity commented:
@yaronkaikov please set up the pgo refresh process once this promotes.
